pkg/snapshot/restorer/restorer.go

```diff
@@ -317,24 +326,260 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
 	return e, nil
 }
 
-// applyDeltaSnapshot applies the events from time sorted list of delta snapshot to etcd sequentially
-func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, snapList snapstore.SnapList) error {
+// applyDeltaSnapshots fetches the events from delta snapshots in parallel and applies them to the embedded etcd sequentially
+func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, ro RestoreOptions) error {
+	snapList := ro.DeltaSnapList
+	numMaxFetchers := ro.MaxFetchers
+
+	// no delta snapshots available
+	if len(snapList) == 0 {
+		return nil
+	}
+
 	firstDeltaSnap := snapList[0]
+
+	r.logger.Infof("Applying delta snapshot %s", path.Join(firstDeltaSnap.SnapDir, firstDeltaSnap.SnapName))
 	if err := r.applyFirstDeltaSnapshot(client, *firstDeltaSnap); err != nil {
 		return err
 	}
 	if err := verifySnapshotRevision(client, snapList[0]); err != nil {
 		return err
 	}
-	for _, snap := range snapList[1:] {
-		if err := r.applyDeltaSnapshot(client, *snap); err != nil {
-			return err
+
+	// no more delta snapshots available
+	if len(snapList) == 1 {
+		return nil
+	}
+
+	var (
+		remainingSnaps         = snapList[1:]
+		numSnaps               = len(remainingSnaps)
+		snapLocations          = make([]string, numSnaps)
+		errCh                  = make(chan error, 1)
+		snapInfoCh             = make(chan snapInfo, numSnaps)
```

Comment:
Keep the `snapInfoCh, fetchedEventsInfoCh,applyInfoCh` channel length to `numFetchers`.

Reply:
`snapInfoCh` needs to have buffer length of `numSnaps`, as we don't want the main function `applyDeltaSnapshots` to wait for fetchers, because in case of failure of one of the fetchers, `applyDeltaSnapshots` will not be able to receive on the `errCh`, as it is still waiting to send on `snapInfoCh` but fetcher(s) have exited.

In case `persister` is taking longer than fetchers, fetchers would have to wait to send on `fetchedEventsInfoCh` if there isn't enough buffer for it. Hence, buffer length is `numSnaps`.

The same logic applies to `applyInfoCh` as well, and usually applying snaps takes longer than persisting them, and we don't want persister to wait for applier. Hence, buffer length is `numSnaps`.